include/seastar/core/memory.hh

```diff
@@ -382,29 +383,76 @@ public:
     void operator=(scoped_large_allocation_warning_disable&&) = delete;
 };
 
-/// Enable/disable heap profiling.
+/// Describes an allocation location in the code.  The location is identified by
+/// its backtrace. One allocation_site can represent many allocations at the
+/// same location. `count` and `size` represent the cumulative sum of all
+/// allocations at the location.
+struct allocation_site {
+    mutable size_t count = 0; // number of live objects allocated at backtrace.
+    mutable size_t size = 0; // amount of bytes in live objects allocated at backtrace.
+    simple_backtrace backtrace; // call site for this allocation
+
+    // All allocation sites are linked to each other. This can be used for easy
+    // iteration across them in gdb scripts where it's difficult to work with
+    // the C++ data structures.
+    mutable const allocation_site* next = nullptr; // next allocation site in the chain
+    mutable const allocation_site* prev = nullptr; // previous allocation site in the chain
+
+    bool operator==(const allocation_site& o) const {
+        return backtrace == o.backtrace;
+    }
+
+    bool operator!=(const allocation_site& o) const {
+        return !(*this == o);
+    }
+};
+
+/// If memory sampling is on returns the current sampled memory live set
+std::vector<allocation_site> sampled_memory_profile();
+
+/// Copies the current sampled set of allocation_sites in the output parameter
+/// (up to size of output vector) Returns amount of copied elements. Useful if
+/// one wants to avoid allocating an output vector (e.g.: under OOM conditions)
+size_t sampled_memory_profile(std::vector<allocation_site>&);
```

Comment:
`std::array` might also be a good match for this kind of thing (since std::array does not inherently require any allocations, whereas vector does)

Reply:
Good idea, though I guess you lock your API into a certain interface then.

Could make it a template I guess but that would mean exposing/moving a lot of the memory.cc private stuff into the header.